### DISCUSSION TITLE
After Upgrading 1.3.0, I have octoprint connection issues..

### DISCUSSION THREAD

**Discussion by thanatas** [Jyers/Marlin#406] [Q&A] [answered]
I have octoprint connection issues. Is anyone having similar issues, I could not figure out it is related to new, Or I will look into octopi? That is why asking other people.

Thanks.

**Comment by smooreace** [comment]
Ive been having issues too, but havent spent any time narrowing down reasons

**Reply by 100s99s** [reply]
I meet same issue, and I disable this.
Marlin\src\gcode\queue.cpp
void GCodeQueue::get_serial_commands() { 
...
/if (c < 0) {
//        // This should never happen, let's log it
//        PORT_REDIRECT(SERIAL_PORTMASK(p));     // Reply to the serial port that sent the command
//        // Crash here to get more information why it failed
//        BUG_ON("SP available but read -1");
//        SERIAL_ERROR_MSG(STR_ERR_SERIAL_MISMATCH);
//        SERIAL_FLUSH();
//        continue;
//}
if(c<0)  continue;

**Reply by aryajuanda** [reply]
i just Disabled #define MEATPACK_ON_SERIAL_PORT_1 on Configuration_adv.h to fix the issue (1.2.2 doesn't include it ) without changing any marlin source code, only configuration file..

i m not using meatpack plugin on my octoprint, so why should it be enabled...

**Reply by 100s99s** [reply]
I compare to the regular marlin code, it doesn't check (c<0).

**Comment by Jyers** [comment]
What kind of issues? Can't connect or intermittent failures?

**Reply by thanatas** [reply]
In first attempt, octo-print could not connect. I tried to again and again. It connects at rate of 57K (not 115). When you power off printer and open again it connects sometimes. I could not solve the pattern.  I haven't got any change to look at all logs.  Another info is that I printed a xyz-cube from octopi, it made some layer problem at the start, I mean, the skirt and extrusion problem. Then I thought that time, it should be related to connection issue because I just printed an xyz-cube before update.

**Reply by joeclifford** [reply]
Are you able to provide your octoprint.log?

And provide a list of octoprint plugins you have enabled?

I was having serial issues with a combination of certain plugins and the emergency parser being enabled.

I managed to resolve by disabling meatpack as it doesn't play nice with arc welder; I never got meatpack working reliably and choose to increase my baudrate instead.

**Reply by joeclifford** [reply]
Do the connection issue go away if you run OctoPrint in safemode?

**Reply by thanatas** [reply]
First attempt;
![image](https://user-images.githubusercontent.com/1278648/114004789-a7d8f880-9867-11eb-973b-13104d35c61e.png)
Second attempt;
![image](https://user-images.githubusercontent.com/1278648/114004947-c939e480-9867-11eb-9394-824934cd1309.png)

And the logs file ofcourse, you will see the plugins. 
[octoprint (4).log](https://github.com/Jyers/Marlin/files/6277672/octoprint.4.log)

You will see the list of plugins in the logs. I am currently using all with previous version (1.2.2)

You will see ngrok's errors, it is negligible, I think.

**Reply by joeclifford** [reply]
Looks like the same symptoms I was seeing. 

Seeing a lot of "Received an error from the printer's firmware: Serial status mismatch"

I thought it was limited to my setup as I self compile with extra features enabled. 

If you are happy compiling Marlin yourself, would you be able to do so with the following changes and test?

1. Disable `#define EMERGENCY_PARSER`. Then compile and test.
2. Disable `#define MEATPACK_ON_SERIAL_PORT_1`. Then compile and test.

Both of these feature changes to how the serial data is handled.

If not, let me know which bin file you have flashed and I can compile the same versions with those features disabled.

Also, does OctoPrint safe mode see the same behaviour?

**Reply by thanatas** [reply]
I will try safe mode and write down the results here. I haven't build marlin for ender 3 v2 yet. I appreciated if you build for me. I am using 

BL Touch + HS + 5x5.4.2.2;

Thanks,

**Reply by joeclifford** [reply]
Zip file contains 3 bin files, 1st is stock 5x5 HS config with latest update, 2nd is with emergency parser disabled, 3rd is meatpack disabled.

[Ender 3 V2.zip](https://github.com/Jyers/Marlin/files/6278254/Ender.3.V2.zip)

**Comment by smooreace** [comment]
Im getting the same when I connect OctoPrint to the printer via in app
"connect " button - "Error reported by printer" on the Pi. and the printer
shows "Printer Kill Reason:  M112 Shutdown    Restart Required"

**Reply by njmV** [reply]
Had the same issue. Recompiled with after commenting // #define MEATPACK_ON_SERIAL_PORT_1 on Configuration_adv.h.

Seems stable now.
No more "Printer Kill Reason:  M112 Shutdown Restart Required" with OcoPi

**Comment by smooreace** [comment]
Flashed with firmware with both edits JoeClifford mentioned worked. Will go
back and make one at a time and report back
"Disable #define EMERGENCY_PARSER. Then compile and test
Disable '#define MEATPACK_ON_SERIAL_PORT_1`. Then compile and test."

**Reply by thanatas** [reply]
I checked all of them,
Two of them failed. Just **firmware-20210408-123037-5x5-no-emergency-meatpack.bin** worked. I will try further and try on printing.

**Reply by joeclifford** [reply]
So I made a mistake in the naming of firmware-20210408-123037-5x5-no-emergency-meatpack.bin

It should have read `firmware-20210408-123037-5x5-no-meatpack.bin`

So it sounds like disabling meatpack solves the issue.

If that is the case, an issue will need to be opened in he main marlin repo and/or the meatpack repo.

**Reply by thanatas** [reply]
I think it is better to test by other peoples, then it will open. Thanks jeo, cheers 👍

**Reply by darco415** [reply]
Meatpack.Parser, and Arc Welder all working here on latest compiled fw. My BLTouch is split off to zmin probe. Also lower baud rate on connection...I usually have mine set 115000 but when it errored, I changed to auto and connected fine...ok it looks like I did build the latest fw but actually didnt load it,,,,so yes I am having connection issues w/parser enabled...reverted back to previous build

**Comment by smooreace** [comment]
I know at least one person running the same custom firmware file that I was
running that IS NOT having a connection issue. To I wouldn't quite say that
this is a concrete solution or even a global issue. Agree that far more
testing is needed

**Reply by thanatas** [reply]
The firmware behaves a bit weird, It crashes to y stops unexpectedly, sometimes it could not print xyz cubes via octoprint.
I returned to 1.2.2 which is stable for octoprint now. I did not any sd card print. It must be test.

**Reply by Jyers** [reply]
I would be interested to know more about how it ended up crashing into the y stop and what went wrong with the cube.

**Reply by thanatas** [reply]
Let me clear, I tried joeclifford new builds. Those builds crashed to y stop.  In spite of the trigger y-stop switch, It tried to continue to go more -y.  

It could not print cube. When I  returned the 1.2.2 it can. But this is again joeclifford' builds issue.  Your (jyers's) builds have no such issue.

Sometimes, we skip more clear info to the developers. Sorry for inconvenience. Thanks for all reply.

**Reply by thanatas** [reply]
> I would be interested to know more about how it ended up crashing into the y stop and what went wrong with the cube.

I figured out the axis problem, I mentioned wrong it is x axis. It crashed left site. In 1.2.2 am using  x= -42.0, y= -12 here x is not worked. I am adding a foto for crash point. 

![image](https://user-images.githubusercontent.com/1278648/114215659-86166900-996e-11eb-8dc2-dd4886382e3e.png)

It tries to go beyond the bed.

**Reply by Jyers** [reply]
This should have been fixed in 1.3.0, is it still happening in this release?

**Reply by thanatas** [reply]
I never had issue 1.2.2 but in 1.3.0 I have problem now.

**Reply by Jyers** [reply]
Oh is that roller assembly on the right part of a mod? In 1.3.0 I added an extended max x position to allow a greater probing region, but it is designed in order to not hit the stock setup. If you have a different assembly there that would explain the crashing. If you compile your own version you simply need to change the max x position in configuration.h

**Reply by aryajuanda** [reply]
![image](https://user-images.githubusercontent.com/4106760/114260890-bbb86200-9a01-11eb-845b-6366df960752.jpeg)

Did you flip your X holder on the right side? 

Or you asemble it incorectly?

Because my ender 3 v2 can go beyond 245 mm om x axis..

**Reply by thanatas** [reply]
Ohh, what have I done!!!

I think, last time I miss that detail, I am changing now, thanks mate. You are great....

**Comment by Kamaris521** [comment]
Upgraded to 1.3.0 today and have had to restart printer several times already with the aforementioned error:
OctoPrint shows "Error reported by printer". E3v2 LCD shows "Printer Kill Reason: M112 Shutdown Restart Required"

**Reply by Kamaris521** [reply]
I was able to use these steps to finally get connected to the printer with OctoPrint:

Settings -> Serial Connection -> Behavior -> Error Handling -> What to do on firmware error
Cancel any ongoing prints but stay connected to the printer
(referenced from https://community.octoprint.org/t/eeprom-crc-mismatch/1998/10)

Additionally, I had just completed and saved a new UBL mesh to eeprom when this error started presenting itself.

**Reply by Jyers** [reply]
I might have to add that to the octoprint settings wiki page if it turns out to be the main problem. I knew something in there was sending an M112, just didn't know why. Thanks so much!

**Reply by Kamaris521** [reply]
Many thanks to you for a wonderful contribution to the community.

**Reply by mincaeuro** [reply]
> I was able to use these steps to finally get connected to the printer with OctoPrint:
> 
> Settings -> Serial Connection -> Behavior -> Error Handling -> What to do on firmware error
> Cancel any ongoing prints but stay connected to the printer
> (referenced from https://community.octoprint.org/t/eeprom-crc-mismatch/1998/10)
> 
> Additionally, I had just completed and saved a new UBL mesh to eeprom when this error started presenting itself.

had the same issues after upgrading, this helped thanks

**Comment by cruzer619** [comment]
question do you have PSU plugin?? I had the same issue.. but once I changed my "post on delay" to 4 sec, Before it was set to 2 seconds.  Everything works as expected.  I'm using 1.3.0 BLtouch 5x5 firmware.
![image](https://user-images.githubusercontent.com/26581299/114110045-71ad7000-988b-11eb-9ed2-cc0b52a3f34f.png)

**Reply by talexop** [reply]
I am using also the PSU plugin but I have to set the delay over 30s for the connection to be without any errors...

**Comment by AnotherGuy22** [comment]
I had the same issue pretty consistently with the M112 Error when waking up OctoDash and starting the serial connection:

Firmware: Jyers/Marlin/Ender_3_V2_Extensible_UI/Current
OctoPrint 1.5.3 
- Bed Leveling Wizard (0.2.4)
- BLTouch Plugin (0.3.4)
- Creality-2x-temperature-reporting-fix (0.0.4)
- Dashboard (1.18.1)
- DisplayLayerProgress Plugin (1.26.0)
- GCODE System Commands (1.0.1)
- Preheat Button (0.7.2)
- PrintTimeGenius Plugin (2.2.7)
- Z Probe Offset (0.5)

OctoDash 2.1.2

**Log:**
2021-04-08 15:16:55,354 - octoprint.util.comm - INFO - Changing monitoring state from "Offline" to "Opening serial connection"
2021-04-08 15:16:55,355 - octoprint.util.comm - INFO - Connecting to port /dev/ttyUSB0, baudrate 115200
2021-04-08 15:16:55,392 - octoprint.util.comm - INFO - Changing monitoring state from "Opening serial connection" to "Connecting"
2021-04-08 15:16:55,396 - octoprint.util.comm - INFO - M110 detected, setting current line number to 0
2021-04-08 15:16:55,400 - octoprint.util.comm - WARNING - Received an error from the printer's firmware: Serial status mismatch
| Last lines in terminal:
| Changing monitoring state from "Offline" to "Opening serial connection"
| Connecting to port /dev/ttyUSB0, baudrate 115200
| Changing monitoring state from "Opening serial connection" to "Connecting"
| Connected to: Serial<id=0xaac15270, open=True>(port='/dev/ttyUSB0', baudrate=115200, bytesize=8, parity='N', stopbits=1, timeout=10.0, xonxoff=False, rtscts=False, dsrdtr=False), starting monitor
| Recv: iXÑaWÍ [Z®+Ñ¡)Error:Serial status mismatch
| Recv: Error:Serial status mismatch
2021-04-08 15:16:55,402 - octoprint.util.comm - INFO - Changing monitoring state from "Connecting" to "Error: Serial status mismatch"
2021-04-08 15:16:55,403 - octoprint.util.comm - INFO - Force-sending M112 to the printer
2021-04-08 15:16:55,416 - octoprint.util.comm - INFO - Changing monitoring state from "Error: Serial status mismatch" to "Offline (Error: Serial status mismatch)"

**Disable '#define MEATPACK_ON_SERIAL_PORT_1` seems to have fixed.**
Btw, love the firmware. Thanks!

**Reply by talexop** [reply]
I can also verify the same behaviour... With 1.3.0 update the problem occurs. 
If it connects the first time then it has no problem to connect again. 
Changing the octoprint serial behaviour on error seems to fix it for now.

**Reply by lloydbigoil** [reply]
may i ask exactly where i need to comment out the meatpack? and is this before compiling to .bin? is anyone compiling jyers with this default commented out?

**Reply by Cal-LF** [reply]
In config_adv.h then compile .bin

**Reply by lloydbigoil** [reply]
down the firmware compiling rabbit hole i go!!!

**Comment by aryajuanda** [comment]
i got similar problem with 1.3.0 
when octo print is up and then i turn on the printer and try to connect its get "Error: Serial status mismatch" and 112 trigered.. (after success connection before or upgrade firmware) if it try to restart printer manually, it still cannot connect until i replug the usb or do some restart before to connect

the easiest way to connect is restart the printer and restart octoprint, its connect without any problem.. or if we just turn on on toprinnt and then turn on the printer it will success to connect without a problem... 

the problem is my raspberry pi is 24 hour service while the printer is not, will try to recompile the firmware ... love some new improvement on 1.3.0 and i don't want to go back to 1.2.2... 

PS : Disable #define MEATPACK_ON_SERIAL_PORT_1 fixed for me

**Reply by thanatas** [reply]
I compiled base marlin before. unfortunately, I am not pretty sure each concepts using to compile this fork with platform io. 

Is there any documentation how to build Jyers's fork for 3x3 or 5x5  or etc. Could you give the info please, I want to test myself. 

Thanks..

**Reply by aryajuanda** [reply]
you can download the source (i use release zip version and extract), download visuat studio code + Plugin (platformio and marlin auto build)
open marlin folder in visual studio code and replace configuration file inside Marlin Folder
file need to change :
Configuration.h
Configuration_adv.h

you can get this file from sample inside Folder Configuration Folder...  

there a lot of tutorial on how to build marlin firmware on the internet.. google is your friend

**Reply by thanatas** [reply]
Yes, I know how to do it. Just confused is there any configuration that is automatically build each output for exam bl-touch-3x3, I need to change manually or any config switch to do that?

**Reply by aryajuanda** [reply]
i saw jyers using a workflow on github to create automatic for each build.. you can check on action...  may be you can fork and run in on your github with your own workflow... i  haven't try it before..

**Reply by Jyers** [reply]
You might have a hard time reworking my workflow for individual use. It utilizes configuration matrices to generate the different skews and is hooked into the GitHub release system. It automatically builds when a new release is published and uploads the files directly to the release assets. Might be more effort than its worth to build one config.

**Reply by thanatas** [reply]
I know how hard to run cd-ci, I set up more than one. I will try it manual ways than. Thank you very much for your valuable comments.

**Comment by franknepomuceno** [comment]
I was having the same issues after the update.
I think I solved it by NOT choosing AUTO on the serial port.
I choosing the serial port as /dev/ttyUSB0

Hope that helps

**Comment by robertluckett** [comment]
It seems that since V1.3, the controller board serial port isn't stable or ready for a short while after startup. I'm also powering up via Octoprint with a TP Link smartplug. Not sure whether this is specific to mesh levelling versions, as I added a BLTouch at the same time as I flashed 1.3 and therefore changed the to the appropriate version for a 4.2.2 board.

From post above, but I see similar in my logs.

> Recv: iX��ÑaWÍ [Z®+Ñ�¡)Error:Serial status mismatch
> Recv: Error:Serial status mismatch

So Octoprint fails safe by sending M112. 

There are serial timeout settings in Octoprint, I tried increasing "Initial baudrate detection pause" but it doesn't seem to introduce a delay in these circumstances.

As a workaround I've set a low default baud rate which won't connect, waiting a short while then manually connecting at 115200.

@Jyers - a cracking piece of work BTW

**Comment by Xploder** [comment]
Same problem here and compiling official Marlin bugfix branch doesn't solve the issue either (at least as of 2 weeks ago), so the issue is upstream.

**Comment by Jyers** [comment, marked as answer]
**Here are the major conclusions we have come to so far**
* As @Xploder said, this seems to stem from upstream
* @franknepomuceno is right that setting the serial port seems to fix the issue, but octoprint does not always give the option for USB
* @Kamaris521 found a great solution to stop octoprint sending M112:
>Settings -> Serial Connection -> Behavior -> Error Handling -> What to do on firmware error
Cancel any ongoing prints but stay connected to the printer

Also deselect "Send M112 on disconnect due to error"

**From these answers we can at least:**
* Prevent the need to restart the printer
* Begin to look into what is really causing this

***

I'm going to mark this as the answer so it will be highlighted and will update it as we learn new information about this issue

**Reply by joeclifford** [reply]
I was going to be creating issues in the main github repo and meatpack repo this everning, will link to the issues here once I have done so.

Creating one for meatpack because disabling meatpack in the config before build also resolves the issue.

**Reply by tubalainen** [reply]
(Qualified) guess: The majority of the users using this firmware are not interested in Meatpack. It has since the release on day one caused strange behaviors and caused issues when included in the Marlin firmware.

FW enabled Meatpack - Not installed Meatpack plugin in Octoprint - **FAIL** (random errors)
Not FW enabled Meatpack - Installed Meatpack plugin in Octoprint - **FAIL** (cannot connect)
FW enabled Meatpack - Installed Meatpack plugin in Octoprint - **Works** (if FW and Plugin versions align)

My pea brain conclusion is that this isn't an option in the firmware that should be default.

**Reply by joeclifford** [reply]
This should probably be moved to its own Ideas discussion.

**Reply by Jyers** [reply]
@tubalainen I would completely agree with your point, if I could recreate that behavior. However in my testing the only connection errors I've gotten have been with the plugin installed and meatpack enabled. In which case it requires connecting twice and then it is fine. When I tested without the plugin, but still having it enabled in the firmware, it behaved perfectly. I could not manage to get the connection to fail. If there is something I am doing which is incorrect, by all means let me know. Before testing also make sure you have all the settings I describe above set in octoprint..

**Comment by asychev** [comment]
With the firmware 1.3.1 and latest Octoprint on RaspberryPI 3B  - no connection at all. Tried different cables, no luck. On PI lsusb does not even show USB device.

**Reply by Jyers** [reply]
So your just getting a no candidates error?

**Reply by asychev** [reply]
> So your just getting a no candidates error?
> 

Sorry, my bad, it was just bad cable (well, actually, 3 of them). Now it works with 115200. Got M112 error once.

**Comment by smooreace** [comment]
All you have to do is what the WIKI tells you to do in Octoprint. And 1.3.0
is REALLY old!

**Comment by smooreace** [comment]
Its nothing more than editing a text file. There are countless how-to's out
there for manipulating every common aspect of Marlin firmware too
